text/0074-global-init.md

```diff
@@ -0,0 +1,176 @@
+OpenTelemetry is designed to support "zero touch" instrumentation for
+third party libraries through the use of a global Tracer and Meter
+provider (factory) instances.  This OTEP proposes a specification for how to
+initialize the global provider (factory) instances.
+
+In some programming environments, it is possible for libraries of code
+to auto-initialize static variables, allowing them to begin operation
+concurrently with the main program, while initializing static program
+state.  This presents a set of opposing requirements: (1) the API
+supports a configurable SDK; (2) third party libraries can use
+OpenTelemetry before the SDK is configured.
+
+The current specification discusses a global provider (factory) for
+named Tracers and Meters as well as a global Propagators instance for
+injectors and extractors, but does not discuss how these are
+initialized or whether their values can be modified during the process
+lifetime.
+
+Global variables face significant opposition from some developers,
+which forces the question: "why support globals at all?".  In
+languages with automatic dependency injection support, then
+conceivably we do not need global variables.  In languages without
+automatic dependency injection, without globals we could not have the
+"zero touch" instrumentation [given as a requirement for the
+project](https://github.com/open-telemetry/oteps/blob/master/text/0001-telemetry-without-manual-instrumentation.md).
+If a third-party library is to be instrumented and integrated without
+modification into an application, either dependency injection or
+global variables are the solution.
+
+Global variables **are a dangerous programming pattern**, but they
+also enable easy integration in languages without automatic dependency
+injection.  To address this risk, this proposal specifies strict
+limits on their initialization.  We propose at-most-once
+initialization for the three global variables in OpenTelemetry.
+Specifically, this specification says that the global Tracer provider,
+the global Meter provider, and the global Propagators instance can
+only be initialized once per process lifetime (except possibly in
+test-only scenarios).
+
+This OTEP also specifies the behavior of the global instances when
+they are used before the SDK is configured and installed, in case this
+cannot be performed by automatic dependency injection.
+
+## Explanation
+
+There are two acceptable ways to provide default instances in
+OpenTelemetry: (1) through dependency injection, (2) through global
+variables initialized at most once.  The feasibility of each approach
+varies by language.  The implementation MUST select one of the
+following strategies.
+
+### Service provider mechanism
+
+Where the language provides a commonly accepted way to inject SDK
+components, it should be preferred.  The Java SPI (Service Provider
+Interface) supports loading and configuring the SDK before it is first
```

Comment:
Note: this probably won't work even in Java in general. Application servers often host many "applications" all running in the same java process. Each would need to be able to have an independently initialized SDK. And, the underlying application server might also need one of it's own. Classloader isolation could solve some of this, but it's definitely a use case that could cause troubles, especially in a java-agent auto-instrumentation scenario.

Reply:
I more-or-less intended to describe what's needed in a language without DI support. @bogdandrutu convinced me that the Java SPI would take care of the Java case, so I put this paragraph in as a stub. I'd be happy to remove this discussion or strengthen it. 

I had assumed that the DI support could handle initializing SDK(s) and injecting them the way you describe--I'm not very familiar with Java DI though.

Reply:
Yes, I'm not convinced that SPI is going to be flexible enough for all the myriad Java use-cases, so I'd definitely weaken this statement. Because the Java ecosystem is so varied, we'll need something very flexible. I'd be willing to bet that for languages/communities that are more opinionated will need less flexibility that Java.

Reply:
> Note: this probably won't work even in Java in general. 

I don't buy this. (a) it worked in OpenTracing. (b) SPI is a widely used pattern in Java, not at all specific to OpenTelemetry; if it was so flawed as the comment implies, everything else in Java ecosystem wouldn't work (even basic logging). And yes, app servers (a legacy notion in itself for almost 10yrs) still handle this perfectly fine with "classloader per application" (which is, incidentally, a totally self-inflicted problem).

Reply:
this is fair, but I want to make sure this ends up working for the auto-instrumentation agent, also. In my experience, it's actually pretty easy for SPI-like constructs (for logging, in particular) to collide with each other in a complex java agent premain environment.

Reply:
I like the expression "premain environment".  If you're talking about what happens before main(), it's probably complex no matter what the language.